### PR TITLE
Add retry policy class and config for operation retry

### PR DESCRIPTION
**Pull request by xyuanlu**
### Issues

- [X] My PR addresses the following Helix issues and references them in the PR description:
#2237 

### Description

- [X] Here are some details about my PR, including screenshots of any UI changes:
This change adds reconnect policy class and CRUD retry operation for MetaClient. Will add impl for adopting reconnect policy in MetaClient in next PR. (test will be add in that PR as well)
 Reconnect policy : User can define behavior for auto reconnect when connection is expired or lost.
 CRUD retry config: User may config a timeout. MetaClient will reissue CRUD operation if connection is re-established within timeout. 

### Tests

- [X] The following tests are written for this issue:

Test will be added when implementing reconnect policy. Will add e2e test for retey connection and CRUD since CRUD retry waits for reconnection established signal.

- The following is the result of the "mvn test" command on the appropriate module:

(If CI test fails due to known issue, please specify the issue and test PR locally. Then copy & paste the result of "mvn test" to here.)

### Changes that Break Backward Compatibility (Optional)

- My PR contains changes that break backward compatibility or previous assumptions for certain methods or API. They include:

(Consider including all behavior changes for public methods or API. Also include these changes in merge description so that other developers are aware of these changes. This allows them to make relevant code changes in feature branches accounting for the new method/API behavior.)

### Documentation (Optional)

- In case of new functionality, my PR adds documentation in the following wiki page:

(Link the GitHub wiki you added)

### Commits

- My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

File: meta-client/src/main/java/org/apache/helix/metaclient/api/ExponentialBackoffReconnectPolicy.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    _maxBackOffInterval = DEFAULT_MAX_EXP_BACKOFF_RETRY_INTERVAL_MS;
+  }
+
+  public ExponentialBackoffReconnectPolicy(long maxBackOffInterval, long initBackoffInterval) {
+    _maxBackOffInterval = maxBackOffInterval;
+    _initBackoffInterval = initBackoffInterval;
+
+  }
+
+  long getMaxBackOffInterval() {
```
Comment (rahulrane50) [inline meta-client/src/main/java/org/apache/helix/metaclient/api/ExponentialBackoffReconnectPolicy.java:52]: May I ask why we have default scope here for these methods?

File: meta-client/src/main/java/org/apache/helix/metaclient/api/MetaClientReconnectPolicy.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+ * Policy to define client re-establish connection behavior when connection to underlying metadata
+ * store is expired.
+ */
+
+public interface MetaClientReconnectPolicy {
+
+  enum RetryPolicyName {
+    NO_RETRY,
+    EXP_BACKOFF,
+    LINEAR_BACKOFF
```
Comment (rahulrane50) [inline meta-client/src/main/java/org/apache/helix/metaclient/api/MetaClientReconnectPolicy.java:32]: Curious, are we going to implement this one in future PRs?

File: meta-client/src/main/java/org/apache/helix/metaclient/factories/MetaClientConfig.java [hunk 3]
```diff
@@ -123,6 +137,21 @@ public B setConnectionInitTimeoutInMillis(long timeout) {
       return self();
     }
 
+    /**
+     * Set timeout in mm for connection initialization timeout
```
Comment (rahulrane50) [inline meta-client/src/main/java/org/apache/helix/metaclient/factories/MetaClientConfig.java:141]: neat typo : ms

File: meta-client/src/main/java/org/apache/helix/metaclient/factories/MetaClientConfig.java [hunk 4]
```diff
@@ -145,6 +174,18 @@ final B self() {
     }
 
     protected void validate() {
+      if (_metaClientReconnectPolicy == null) {
+        _metaClientReconnectPolicy = new ExponentialBackoffReconnectPolicy();
```
Comment (rahulrane50) [inline meta-client/src/main/java/org/apache/helix/metaclient/factories/MetaClientConfig.java:178]: Neat : looks like we have default retry policy as exponential polity. May be we can comment in this Metaclientretry policy

**Review by rahulrane50 — COMMENTED** [review]
Overall LGTM! One general comment, Do we want to name all class as "*reconnect*policy" or "*retry*policy"? From your description and functionality it looks like later one to me.

**Comment by xyuanlu** [comment]
> Overall LGTM! One general comment, Do we want to name all class as "_reconnect_policy" or "_retry_policy"? From your description and functionality it looks like later one to me.

TFTR. 
Retry connection policy - how we reconnection in a separate thread when connection lost. Different policies can be defined for this.
Retry - how we define behavior for retry CRUD operation when connection is lost. We only define timeout for this.

File: meta-client/src/main/java/org/apache/helix/metaclient/api/MetaClientReconnectPolicy.java (hunk 2 above)
Comment (xyuanlu) [reply meta-client/src/main/java/org/apache/helix/metaclient/api/MetaClientReconnectPolicy.java:32]: EXP_BACKOFF is implemented in ZKClient. We will reuse.
the other two is scheduled for future.

File: meta-client/src/main/java/org/apache/helix/metaclient/factories/MetaClientConfig.java (hunk 3 above)
Comment (xyuanlu) [reply meta-client/src/main/java/org/apache/helix/metaclient/factories/MetaClientConfig.java]: TFTR. Updated.

File: meta-client/src/main/java/org/apache/helix/metaclient/api/MetaClientReconnectPolicy.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
+package org.apache.helix.metaclient.api;
```
Comment (qqu0127) [inline meta-client/src/main/java/org/apache/helix/metaclient/api/MetaClientReconnectPolicy.java:1]: nit on package management. Maybe this deserves a separate module as `retryPolicy`? And usually we should separate its interface and the implementation.

File: meta-client/src/main/java/org/apache/helix/metaclient/api/MetaClientReconnectPolicy.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/**
+ * Policy to define client re-establish connection behavior when connection to underlying metadata
+ * store is expired.
+ */
```
Comment (qqu0127) [inline meta-client/src/main/java/org/apache/helix/metaclient/api/MetaClientReconnectPolicy.java:25]: Not sure if I understand the purpose here. What behavior does this interface abstract? If users want to implement their own policy, how do they do that?

File: meta-client/src/main/java/org/apache/helix/metaclient/factories/MetaClientConfig.java [hunk 7]
```diff
@@ -123,6 +137,27 @@ public B setConnectionInitTimeoutInMillis(long timeout) {
...
+      return self();
+    }
+
+    /**
+     * Set reconnect policy when connection is lost or expired. By default is
+     * ExponentialBackoffReconnectPolicy
+     * @param reconnectPolicy an instance of type MetaClientReconnectPolicy
+     * @return
+     */
+    public B setMetaClientReconnectPolicy(MetaClientReconnectPolicy reconnectPolicy) {
```
Comment (qqu0127) [inline meta-client/src/main/java/org/apache/helix/metaclient/factories/MetaClientConfig.java:156]: If we want to use default value, can we assign it at declaration time?

File: meta-client/src/main/java/org/apache/helix/metaclient/factories/MetaClientConfig.java [hunk 8]
```diff
@@ -32,10 +34,18 @@ public enum StoreType {
   // Wait for init timeout time until connection is initiated
   private final long _connectionInitTimeoutInMillis;
 
+  // Operation failed because of connection lost will be auto retried if connection has recovered
+  // within timeout time.
+  private final long _operationRetryTimeoutInMillis;
+
```
Comment (qqu0127) [inline meta-client/src/main/java/org/apache/helix/metaclient/factories/MetaClientConfig.java:40]: It's a little confusing we have both this one and `MetaClientReconnectPolicy`. Can we combine these two?

File: meta-client/src/main/java/org/apache/helix/metaclient/factories/MetaClientConfig.java (hunk 7 above)
Comment (xyuanlu) [reply meta-client/src/main/java/org/apache/helix/metaclient/factories/MetaClientConfig.java:156]: If we want to assign at declaration time and user also set their own policy, we will be creating an default policy object and then throw away.

File: meta-client/src/main/java/org/apache/helix/metaclient/factories/MetaClientConfig.java (hunk 8 above)
Comment (xyuanlu) [reply meta-client/src/main/java/org/apache/helix/metaclient/factories/MetaClientConfig.java:40]: operationRetryTimeout if for operation retry. Client does not reissue any request until it gets notification that connection is re-established. 
ReconnectPolicy is to define auto reconnect policy when connection to metadata store is lost. 
I think they belongs to different categories.

File: meta-client/src/main/java/org/apache/helix/metaclient/api/MetaClientReconnectPolicy.java (hunk 5 above)
Comment (xyuanlu) [reply meta-client/src/main/java/org/apache/helix/metaclient/api/MetaClientReconnectPolicy.java]: TFTR. Updated.

**Review by rahulrane50 — APPROVED** [review]
Thanks for addressing the comments, LGTM!

**Review by qqu0127 — APPROVED** [review]

**Pull request merged**